### DISCUSSION TITLE
Module not found: Can't resolve 'fs'

### DISCUSSION THREAD

**Discussion by rinarakaki** [vercel/next.js#39549] [Help]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

    Operating System:
      Platform: linux
      Arch: x64
      Version: #1 SMP Thu Mar 17 17:08:06 UTC 2022
    Binaries:
      Node: 18.7.0
      npm: 8.17.0
      Yarn: 1.22.17
      pnpm: N/A
    Relevant packages:
      next: 12.2.4
      eslint-config-next: 12.2.4
      react: 18.2.0
      react-dom: 18.2.0

### Describe the Bug

I'm trying to use a component in the `components` directory that needs to use `fs` module to access local files.

I know I can fix this kind of error if it's a component in the `pages` directory then I use `fs` in the `getStaticProps` function.
My problem is I want to use it from the `components` directory to make it a shared utility.
How can I solve this? 

### Expected Behavior

Use node.js built-in modules such as `fs` and `path` from the `components` directory during SSG.

### Link to reproduction

-

### To Reproduce

-

**Comment by mimiza** [comment]
I am facing the same issue now. I'm using the same next version 12.2.4 and cannot import fs in /pages components

**Reply by zelal-dev** [reply]
I have same issue

**Reply by haobing55** [reply]
I have the same issure,too.

**Reply by abdinurkz** [reply]
also same issue

**Reply by 1503963513** [reply]
I have the same question

**Reply by leottx** [reply]
Same here... It was solved?

**Reply by kuihao** [reply]
same issue 2023/11

**Reply by LaAlquimia** [reply]
2023/12

**Comment by icyJoseph** [comment]
Hi, 

You could tell WebPack to ignore this package.

```js
/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  webpack(config, { nextRuntime }) { 
    // as of Next.js latest versions, the nextRuntime is preferred over `isServer`, because of edge-runtime
    if (typeof nextRuntime === "undefined") {
      const { IgnorePlugin } = require("webpack");
      const ignoreFs = new IgnorePlugin({ resourceRegExp: /fs/ });
      config.plugins.push(ignoreFs);
    }

    return config;
  },
};

module.exports = nextConfig;
```

Here's a read that can explain further, https://arunoda.me/blog/ssr-and-server-only-modules

However, in general it is best to structure your code so that the code elimination tool (demo https://next-code-elimination.vercel.app/), can take care of removing unnecessary packages.

**Reply by neuromantic99** [reply]
Where do you put this code?

**Reply by icyJoseph** [reply]
In `next.config.js` I think one could also do:

```js
if (typeof nextRuntime === "undefined") {
      config.resolve.fallback = {
                ...config.resolve.fallback,
                fs: false,
         };  
}
```

Create one if you don't have it yet, at the root of your project.

**Reply by xuchenyang123** [reply]
This plan is effective

**Reply by Pacifist-Today** [reply]
after using icyJoseph's code, I have this

./node_modules/cross-spawn/index.js:3:0
Module not found: Can't resolve 'child_process'

https://nextjs.org/docs/messages/module-not-found

Import trace for requested module:
./node_modules/@pkgr/utils/lib/index.cjs
./node_modules/synckit/lib/index.cjs
./node_modules/eslint-import-resolver-typescript/lib/index.cjs
./node_modules/eslint-config-next/index.js
./actions/index.js
./pages/index.js

**Reply by mohit3500** [reply]
Mine too, is it resolved? @Pacifist-Today

**Reply by Preet-Sojitra** [reply]
I used this and I am getting following error:
```
error node_modules/pdf2json/lib/pdf.js (66:0) @ eval
error ReferenceError: nodeUtil is not defined
```
That `fs` error is gone. 

Also, I am using external library called `PDFReader`, which was generating that `fs` error .

**Reply by mcewball13** [reply]
Did you figure this out? I'm using PDF renderer and it just started after a yarn upgrade. Now I can't get it working again.

**Comment by thngdude** [comment]
This worked for me @icyJoseph. Thx
